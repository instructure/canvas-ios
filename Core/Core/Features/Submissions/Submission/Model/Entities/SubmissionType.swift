//
// This file is part of Canvas.
// Copyright (C) 2025-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation
import UniformTypeIdentifiers

public enum SubmissionType: String, Codable, CaseIterable {
    case discussion_topic
    case external_tool
    case media_recording
    case none
    case not_graded
    case online_quiz
    case online_text_entry
    case online_upload
    case online_url
    case on_paper
    case basic_lti_launch
    case wiki_page
    case student_annotation

    public var localizedString: String {
        switch self {
        case .discussion_topic:
            return String(localized: "Discussion Comment", bundle: .core)
        case .external_tool, .basic_lti_launch:
            return String(localized: "External Tool", bundle: .core)
        case .media_recording:
            return String(localized: "Media Recording", bundle: .core)
        case .none:
            return String(localized: "No Submission", bundle: .core)
        case .not_graded:
            return String(localized: "Not Graded", bundle: .core)
        case .online_quiz:
            return String(localized: "Quiz", bundle: .core)
        case .online_text_entry:
            return String(localized: "Text Entry", bundle: .core)
        case .online_upload:
            return String(localized: "File Upload", bundle: .core)
        case .online_url:
            return String(localized: "Website URL", bundle: .core)
        case .on_paper:
            return String(localized: "On Paper", bundle: .core)
        case .wiki_page:
            return String(localized: "Page", bundle: .core)
        case .student_annotation:
            return String(localized: "Student Annotation", bundle: .core)

        }
    }
}

extension Array where Element == SubmissionType {
    var isOnline: Bool {
        if contains(.on_paper) || contains(.not_graded) || contains(.none) {
            return false
        }
        return true
    }

    public var allowedMediaTypes: [String] {
        var types  = [UTType.movie.identifier]

        if contains(.media_recording) && !contains(.online_upload) {
            types.append(UTType.audio.identifier)
        } else {
            types.append(UTType.image.identifier)
        }
        return types
    }

    public func allowedUTIs(allowedExtensions: [String] = []) -> [UTI] {
        var utis: [UTI] = []

        if contains(.online_upload) {
            if allowedExtensions.isEmpty {
                utis += [.any]
            } else {
                utis += UTI.from(extensions: allowedExtensions)
            }
        }

        if contains(.media_recording) {
            utis += [.video, .audio]
        }

        if contains(.online_text_entry) {
            utis += [.text]
        }

        if contains(.online_url) {
            utis += [.url]
        }

        return utis
    }

    public func isStudioAccepted(
        allowedExtensions: [String]
    ) -> Bool {
        guard self.contains(.online_upload) else {
            return false
        }

        if allowedExtensions.isEmpty {
            return true
        }

        for allowedExtension in allowedExtensions {
            guard let fileType = UTType(filenameExtension: allowedExtension) else {
                continue
            }
            if fileType.conforms(to: .audiovisualContent) {
                return true
            }
        }

        return false
    }
}
