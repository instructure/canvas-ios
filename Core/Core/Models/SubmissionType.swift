//
// Copyright (C) 2018-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import Foundation

public enum SubmissionType: String, Codable {
    case discussion_topic
    case external_tool
    case media_recording
    case none
    case not_graded
    case online_quiz
    case online_text_entry
    case online_upload
    case online_url
    case on_paper
    case basic_lti_launch

    public var localizedString: String {
        switch self {
        case .discussion_topic:
            return NSLocalizedString("Discussion Comment", bundle: .core, comment: "")
        case .external_tool, .basic_lti_launch:
            return NSLocalizedString("External Tool", bundle: .core, comment: "")
        case .media_recording:
            return NSLocalizedString("Media Recording", bundle: .core, comment: "")
        case .none:
            return NSLocalizedString("No Submission", bundle: .core, comment: "")
        case .not_graded:
            return NSLocalizedString("Not Graded", bundle: .core, comment: "")
        case .online_quiz:
            return NSLocalizedString("Quiz", bundle: .core, comment: "")
        case .online_text_entry:
            return NSLocalizedString("Text Entry", bundle: .core, comment: "")
        case .online_upload:
            return NSLocalizedString("File Upload", bundle: .core, comment: "")
        case .online_url:
            return NSLocalizedString("Website URL", bundle: .core, comment: "")
        case .on_paper:
            return NSLocalizedString("On Paper", bundle: .core, comment: "")
        }
    }
}
