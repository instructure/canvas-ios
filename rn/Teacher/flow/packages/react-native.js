//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
