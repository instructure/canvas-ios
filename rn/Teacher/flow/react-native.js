//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
