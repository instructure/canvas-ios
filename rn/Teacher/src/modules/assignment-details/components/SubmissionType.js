//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/**
 * @flow
 */

import React, { Component } from 'react'
import { Text } from '../../../common/text'
import { submissionTypes } from '../../../common/submissionTypes'

import {
  View,
  StyleSheet,
} from 'react-native'

export default class SubmissionType extends Component<Object> {
  render () {
    if (!this.props.data) return null
    const types = submissionTypes()
    return (<View style={style.vertical}>
      {this.props.data.map(item =>
        <Text style={{ flex: 1 }} key={item} testID='assignment-details.submission-type.details-lbl'>{types[item]}</Text>
      )}
    </View>)
  }
}

const style = StyleSheet.create({
  vertical: {
    flexDirection: 'column',
  },
})
