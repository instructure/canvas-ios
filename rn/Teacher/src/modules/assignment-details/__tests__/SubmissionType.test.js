//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/**
 * @flow
 */

import 'react-native'
import React from 'react'
import SubmissionType from '../components/SubmissionType'
import renderer from 'react-test-renderer'

const defaultProps = {
  data: ['foo', 'bar'],
}

test('render', () => {
  let tree = renderer.create(
    <SubmissionType {...defaultProps} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('render 0 types', () => {
  defaultProps.data = []
  let tree = renderer.create(
    <SubmissionType {...defaultProps} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('render single type', () => {
  defaultProps.data = ['foo']
  let tree = renderer.create(
    <SubmissionType {...defaultProps} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('render no data', () => {
  let props = {
    ...defaultProps,
    data: null,
  }
  let tree = renderer.create(
    <SubmissionType {...props} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
