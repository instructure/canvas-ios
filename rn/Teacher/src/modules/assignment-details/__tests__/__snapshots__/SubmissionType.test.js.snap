// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 0 types 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
/>
`;

exports[`render 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2D3B45",
          "fontFamily": "Lato-Regular",
          "fontSize": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="assignment-details.submission-type.details-lbl"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#2D3B45",
          "fontFamily": "Lato-Regular",
          "fontSize": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="assignment-details.submission-type.details-lbl"
  />
</View>
`;

exports[`render no data 1`] = `null`;

exports[`render single type 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2D3B45",
          "fontFamily": "Lato-Regular",
          "fontSize": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="assignment-details.submission-type.details-lbl"
  />
</View>
`;
